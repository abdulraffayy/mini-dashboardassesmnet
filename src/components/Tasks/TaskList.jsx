/**
 * TaskList Component
 * 
 * Provides a drag-and-drop task management interface with the following features:
 * - Add new tasks
 * - Toggle task completion
 * - Remove tasks
 * - Reorder tasks via drag and drop
 * - Persists changes to storage
 * - Integrates with notifications for task actions
 */
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
    DndContext,
    closestCenter,
    KeyboardSensor,
    PointerSensor,
    useSensor,
    useSensors,
} from '@dnd-kit/core';
import {
    arrayMove,
    SortableContext,
    sortableKeyboardCoordinates,
    verticalListSortingStrategy,
} from '@dnd-kit/sortable';
import { useSortable } from '@dnd-kit/sortable';
import { CSS } from '@dnd-kit/utilities';
import {
    selectAllTasks,
    addTaskWithNotification,
    toggleTaskWithNotification,
    removeTaskWithNotification,
    reorderTasks
} from '../../features/tasks/tasksSlice';

const DragHandle = () => (
  <svg className="w-8 h-8 text-gray-300" fill="none" viewBox="0 0 24 24">
    <rect x="6" y="7" width="12" height="2.2" rx="1" fill="currentColor"/>
    <rect x="6" y="11" width="12" height="2.2" rx="1" fill="currentColor"/>
    <rect x="6" y="15" width="12" height="2.2" rx="1" fill="currentColor"/>
  </svg>
);

const CustomCheckbox = ({ checked, onChange }) => (
  <span className="inline-flex items-center justify-center">
    <input
      type="checkbox"
      checked={checked}
      onChange={onChange}
      className="appearance-none h-6 w-6 border-2 border-blue-500 bg-white checked:bg-blue-500 checked:border-blue-500 focus:ring-0 cursor-pointer"
      style={{ borderRadius: 0 }}
    />
    {checked && (
      <svg className="absolute w-4 h-4 text-white pointer-events-none" viewBox="0 0 20 20" fill="none">
        <polyline points="5 11 9 15 15 7" stroke="white" strokeWidth="2.2" strokeLinecap="round" strokeLinejoin="round" />
      </svg>
    )}
  </span>
);

const DEFAULT_TASKS = [
  { id: '1', text: 'Finish report', completed: true },
  { id: '2', text: 'Email client', completed: false },
  { id: '3', text: 'Update website', completed: false },
  { id: '4', text: 'Draft presentation', completed: false },
];

/**
 * SortableTask Component
 * Individual task item that can be dragged and sorted
 * 
 * @param {Object} props
 * @param {number} props.id - Task ID
 * @param {string} props.text - Task text
 * @param {boolean} props.completed - Task completion status
 * @param {Function} props.onToggle - Callback for toggling completion
 */
const SortableTask = ({ id, text, completed, onToggle, isLast }) => {
    const {
        attributes,
        listeners,
        setNodeRef,
        transform,
        transition,
    } = useSortable({ id });

    const style = {
        transform: CSS.Transform.toString(transform),
        transition,
    };

    return (
        <li
            ref={setNodeRef}
            style={style}
            {...attributes}
            className={`flex items-center h-14 px-4 ${!isLast ? 'border-b border-gray-200' : ''}`}
        >
            <CustomCheckbox checked={completed} onChange={() => onToggle(id, text)} />
            <span className="ml-4 text-lg font-semibold text-gray-900 select-none flex-1">{text}</span>
            <span {...listeners} className="cursor-grab ml-4 flex items-center">
              <DragHandle />
            </span>
        </li>
    );
};

export default function TaskList() {
    const tasksFromStore = useSelector(selectAllTasks);
    const tasks = tasksFromStore && tasksFromStore.length > 0 ? tasksFromStore : DEFAULT_TASKS;
    const dispatch = useDispatch();

    // Configure DnD sensors
    const sensors = useSensors(
        useSensor(PointerSensor),
        useSensor(KeyboardSensor, {
            coordinateGetter: sortableKeyboardCoordinates,
        })
    );

    /**
     * Handles drag end event to reorder tasks
     * @param {Object} event - DnD event object
     */
    const handleDragEnd = (event) => {
        const { active, over } = event;

        if (active.id !== over.id) {
            const oldIndex = tasks.findIndex((task) => task.id === active.id);
            const newIndex = tasks.findIndex((task) => task.id === over.id);

            dispatch(reorderTasks({
                sourceIndex: oldIndex,
                destinationIndex: newIndex
            }));
        }
    };

    return (
        <div className="bg-white p-8 rounded-xl shadow-md">
            <h2 className="text-xl font-bold mb-6">Tasks</h2>
            <DndContext
                sensors={sensors}
                collisionDetection={closestCenter}
                onDragEnd={handleDragEnd}
            >
                <SortableContext
                    items={tasks.map(task => task.id)}
                    strategy={verticalListSortingStrategy}
                >
                    <ul>
                        {tasks.map((task, idx) => (
                            <SortableTask
                                key={task.id}
                                id={task.id}
                                text={task.text}
                                completed={task.completed}
                                onToggle={(id, text) => dispatch(toggleTaskWithNotification(id, text))}
                                isLast={idx === tasks.length - 1}
                            />
                        ))}
                    </ul>
                </SortableContext>
            </DndContext>
        </div>
    );
}
