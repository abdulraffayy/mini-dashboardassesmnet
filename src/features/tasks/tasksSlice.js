import { createSlice } from '@reduxjs/toolkit';
import storage from '../../services/storage';
import { addNotification } from '../notifications/notificationSlice';

const DEFAULT_TASKS = [
  { id: '1', text: 'Finish report', completed: true },
  { id: '2', text: 'Email client', completed: false },
  { id: '3', text: 'Update website', completed: false },
];

const storedTasks = storage.getTasks();
const initialState = {
  tasks: storedTasks && storedTasks.length > 0 ? storedTasks : DEFAULT_TASKS
};

const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    addTask: (state, action) => {
      const newTask = {
        id: Date.now(),
        text: action.payload,
        completed: false,
        createdAt: new Date().toISOString()
      };
      state.tasks.push(newTask);
      storage.setTasks(state.tasks);
    },
    toggleTask: (state, action) => {
      const task = state.tasks.find(t => t.id === action.payload);
      if (task) {
        task.completed = !task.completed;
        storage.setTasks(state.tasks);
      }
    },
    removeTask: (state, action) => {
      state.tasks = state.tasks.filter(t => t.id !== action.payload);
      storage.setTasks(state.tasks);
    },
    reorderTasks: (state, action) => {
      const { sourceIndex, destinationIndex } = action.payload;
      const [removed] = state.tasks.splice(sourceIndex, 1);
      state.tasks.splice(destinationIndex, 0, removed);
      storage.setTasks(state.tasks);
    },
    loadTasks: (state) => {
      state.tasks = storage.getTasks();
    }
  }
});

// Action creators
export const { addTask, toggleTask, removeTask, reorderTasks, loadTasks } = tasksSlice.actions;

// Thunks
export const addTaskWithNotification = (text) => (dispatch) => {
  dispatch(addTask(text));
  dispatch(addNotification(`New task added: ${text}`));
};

export const toggleTaskWithNotification = (id, text) => (dispatch) => {
  dispatch(toggleTask(id));
  dispatch(addNotification(`Task completed: ${text}`));
};

export const removeTaskWithNotification = (id, text) => (dispatch) => {
  dispatch(removeTask(id));
  dispatch(addNotification(`Task removed: ${text}`));
};

// Selectors
export const selectAllTasks = state => state.tasks.tasks;
export const selectCompletedTasks = state => state.tasks.tasks.filter(task => task.completed);
export const selectIncompleteTasks = state => state.tasks.tasks.filter(task => !task.completed);

export default tasksSlice.reducer;
