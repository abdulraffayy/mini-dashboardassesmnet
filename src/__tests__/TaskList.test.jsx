import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import TaskList from '../components/Tasks/TaskList';
import tasksSlice from '../features/tasks/tasksSlice';
import notificationsSlice from '../features/notifications/notificationSlice';

// Mock DnD functionality
jest.mock('@dnd-kit/core', () => ({
    ...jest.requireActual('@dnd-kit/core'),
    DndContext: ({ children }) => children,
    useSensor: () => null,
    useSensors: () => null,
}));

jest.mock('@dnd-kit/sortable', () => ({
    ...jest.requireActual('@dnd-kit/sortable'),
    SortableContext: ({ children }) => children,
    useSortable: () => ({
        attributes: {},
        listeners: {},
        setNodeRef: () => { },
        transform: null,
        transition: null,
    }),
}));

describe('TaskList', () => {
    const createStore = (initialState = {}) => {
        return configureStore({
            reducer: {
                tasks: tasksSlice,
                notifications: notificationsSlice
            },
            preloadedState: {
                tasks: {
                    tasks: initialState.tasks || []
                },
                notifications: {
                    items: initialState.notifications || []
                }
            }
        });
    };

    const renderWithStore = (
        ui,
        { initialState = {}, store = createStore(initialState) } = {}
    ) => {
        return {
            ...render(<Provider store={store}>{ui}</Provider>),
            store,
        };
    };

    test('toggles task completion', async () => {
        const initialState = {
            tasks: [
                { id: '1', text: 'Finish report', completed: false }
            ]
        };
        renderWithStore(<TaskList />, { initialState });
        const checkbox = screen.getByRole('checkbox');
        expect(checkbox).not.toBeChecked();
        await userEvent.click(checkbox);
        expect(checkbox).toBeChecked();
    });

    test('matches snapshot', () => {
        const initialState = {
            tasks: [
                { id: '1', text: 'Finish report', completed: true },
                { id: '2', text: 'Email client', completed: false }
            ]
        };
        const { container } = renderWithStore(<TaskList />, { initialState });
        expect(container).toMatchSnapshot();
    });
});
