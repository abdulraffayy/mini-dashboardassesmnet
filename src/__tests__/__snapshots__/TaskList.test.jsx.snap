// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList matches snapshot 1`] = `
<div>
  <div
    class="bg-white p-8 rounded-xl shadow-md"
  >
    <h2
      class="text-xl font-bold mb-6"
    >
      Tasks
    </h2>
    <ul>
      <li
        class="flex items-center h-14 px-4 border-b border-gray-200"
      >
        <span
          class="inline-flex items-center justify-center"
        >
          <input
            checked=""
            class="appearance-none h-6 w-6 border-2 border-blue-500 bg-white checked:bg-blue-500 checked:border-blue-500 focus:ring-0 cursor-pointer"
            style="border-radius: 0;"
            type="checkbox"
          />
          <svg
            class="absolute w-4 h-4 text-white pointer-events-none"
            fill="none"
            viewBox="0 0 20 20"
          >
            <polyline
              points="5 11 9 15 15 7"
              stroke="white"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2.2"
            />
          </svg>
        </span>
        <span
          class="ml-4 text-lg font-semibold text-gray-900 select-none flex-1"
        >
          Finish report
        </span>
        <span
          class="cursor-grab ml-4 flex items-center"
        >
          <svg
            class="w-8 h-8 text-gray-300"
            fill="none"
            viewBox="0 0 24 24"
          >
            <rect
              fill="currentColor"
              height="2.2"
              rx="1"
              width="12"
              x="6"
              y="7"
            />
            <rect
              fill="currentColor"
              height="2.2"
              rx="1"
              width="12"
              x="6"
              y="11"
            />
            <rect
              fill="currentColor"
              height="2.2"
              rx="1"
              width="12"
              x="6"
              y="15"
            />
          </svg>
        </span>
      </li>
      <li
        class="flex items-center h-14 px-4 "
      >
        <span
          class="inline-flex items-center justify-center"
        >
          <input
            class="appearance-none h-6 w-6 border-2 border-blue-500 bg-white checked:bg-blue-500 checked:border-blue-500 focus:ring-0 cursor-pointer"
            style="border-radius: 0;"
            type="checkbox"
          />
        </span>
        <span
          class="ml-4 text-lg font-semibold text-gray-900 select-none flex-1"
        >
          Email client
        </span>
        <span
          class="cursor-grab ml-4 flex items-center"
        >
          <svg
            class="w-8 h-8 text-gray-300"
            fill="none"
            viewBox="0 0 24 24"
          >
            <rect
              fill="currentColor"
              height="2.2"
              rx="1"
              width="12"
              x="6"
              y="7"
            />
            <rect
              fill="currentColor"
              height="2.2"
              rx="1"
              width="12"
              x="6"
              y="11"
            />
            <rect
              fill="currentColor"
              height="2.2"
              rx="1"
              width="12"
              x="6"
              y="15"
            />
          </svg>
        </span>
      </li>
    </ul>
  </div>
</div>
`;
